Update API client to perform retries

colinking opened the pull request:
Updates the API client to perform retries, similar to our other API clients (e.g. https://github.com/airplanedev/js-sdk/blob/main/packages/lib/internal/fetcher.ts). The only differences are (1) the retry policy performs the first retry with no delay (I'll backport this elsewhere) and (2) it doesn't support parsing non-integer `Retry-After` header values. Our API does not use that feature.

I wanted to use the `Retry` adapter from `requests`, but this doesn't let you customize the retry behavior beyond selecting status codes to retry and we have some additional logic involving the `X-Airplane-Retryable` response header.

Resolves DTS-236

colinking commented:
Current dependencies on/for this PR:
* main
  * **PR #80** <a href="https://app.graphite.dev/github/pr/airplanedev/python-sdk/80" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/airplanedev/python-sdk/80?utm_source=stack-comment).

ericliuche commented on airplane/api/client.py:
```diff
@@ -269,18 +261,127 @@ def get_prompt(self, prompt_id: str) -> Dict[str, Any]:
+        return resp["prompt"]
+
+    def __request(
+        self, method: str, path: str, params: dict = None, body: JSONType = None
```
method can be `method: Literal['POST', 'GET', etc]` and literal comes from `from typing_extensions import Literal.

you won't be able to use just `dict` either since that's a new feature not available in `3.7`, so you'll want to use `Dict[str, Any]`. 

also `= None` without `Optional` is inconsistently handled between static type checkers, e.g. mypy will say it's okay but `pylance` will show an error. you want `Optional[Dict[...]]` and `Optional[JSONType]`...

ericliuche commented on airplane/api/client.py:
```diff
@@ -269,18 +261,127 @@ def get_prompt(self, prompt_id: str) -> Dict[str, Any]:
+                    raise APIClient.__http_error_from_resp(resp)
+
+                retry_after_seconds = APIClient.__parse_retry_after(resp)
+            except HTTPError as err:
+                raise err
+            except (
+                requests.exceptions.ConnectTimeout,
+                requests.exceptions.Timeout,
+            ) as err:
+                if retries == max_retries:
+                    raise err
```
nit: dont' need `as err` or `err`,

```
            except (
                requests.exceptions.ConnectTimeout,
                requests.exceptions.Timeout,
            ):
                if retries == max_retries:
                    raise
```

ericliuche commented on airplane/api/client.py:
```diff
@@ -269,18 +261,127 @@ def get_prompt(self, prompt_id: str) -> Dict[str, Any]:
+                    raise APIClient.__http_error_from_resp(resp)
+
+                retry_after_seconds = APIClient.__parse_retry_after(resp)
+            except HTTPError as err:
+                raise err
```
```
            except HTTPError:
                raise
```

ericliuche commented on airplane/api/client.py:
```diff
@@ -269,18 +261,127 @@ def get_prompt(self, prompt_id: str) -> Dict[str, Any]:
+    @classmethod
+    def __is_json_response(cls, resp: Response) -> bool:
+        header = resp.headers.get("content-type")
+        return header is not None and header.find("application/json") >= 0
```
`return resp.headers.get('content-type') == 'application/json'`

also, just move this function out of the class since it's static, have a `is_json()` method that you call

colinking replied:
Just handling `application/json; charset=utf-8`, though our API currently always returns `application/json`

ericliuche replied:
ah okay, then `return  'application/json' in resp.headers.get('content-type', '')` should work too

ericliuche commented on airplane/api/client.py:
```diff
@@ -269,18 +261,127 @@ def get_prompt(self, prompt_id: str) -> Dict[str, Any]:
+
+    @classmethod
+    def __parse_retry_after(cls, resp: Response) -> int:
+        header = resp.headers.get("retry-after")
+        if not header:
+            return 0
+        try:
+            return int(header)
+        except ValueError:
+            return 0
```
same comment as above, move method outside of class

```
retry = resp.headers.get("retry-after", 0)
try:
    return int(retry)
except ValueError:
    return 0

ericliuche commented on airplane/api/client.py:
```diff
@@ -269,18 +261,127 @@ def get_prompt(self, prompt_id: str) -> Dict[str, Any]:
+        except ValueError:
+            return 0
+
+    @classmethod
+    def __compute_retry_delay(cls, retries: int) -> int:
+        if retries <= 1:
+            return 0
+        base_sec = 0.1
+        cap_sec = 30
+        return randrange(0, min(cap_sec, base_sec * 2 ** (retries - 1)))
```
move out

colinking commented:
Alright @ericliuche this is ready for another review. I've gone back and added tests for the new client behavior. I also updated existing tests that previously depended on the underlying client's behavior.

colinking commented on airplane/api/client.py:
```diff
@@ -22,22 +26,23 @@ class ClientOpts:
+    team_id: str
+    run_id: str
+    tunnel_token: str
+    sandbox_token: str
```
Should I add defaults for the new fields in case folks are directly initializing an `APIClient`?
